### PR TITLE
Use link caching for $management operations (renew locks)

### PR DESCRIPTION
**Pull request by richardpark-msft**
@serbrech found a bug where we appear to be having issues when creating a large # of links (in his testing it's somewhere around 200-500). 

Our RPCClient (which is what they were using) created a link for _each_ call made, which added up quickly when you were doing parallel lock renewals.

This change makes it so we cache (when possible) a link per entity/address and re-use it for any other other calls that occur. This change goes hand in hand with the change made in [Azure/azure-amqp-common-go/pull/53](https://github.com/Azure/azure-amqp-common-go/pull/53)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -17,3 +17,8 @@ require (
 	golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd // indirect
 	nhooyr.io/websocket v1.8.6
 )
+
+replace (
```
Comment (jhendrixMSFT) [inline go.mod:21]: This should be removed.

File: rpc.go [hunk 2]
```diff
@@ -39,49 +39,76 @@ import (
...
 func newRPCClient(ctx context.Context, ec entityConnector, opts ...rpcClientOption) (*rpcClient, error) {
 	r := &rpcClient{
-		ec: ec,
+		ec:            ec,
+		linkCache:     map[string]*rpc.Link{},
+		newRPCLink:    rpc.NewLink,
+		newAMQPClient: newAMQPClient,
 	}
 
+	r.closeAMQPClient = func() error { return r.client.Close() }
```
Comment (jhendrixMSFT) [inline rpc.go:75]: You should be able to shorten this to `r.closeAMQPClient = r.client.Close`

File: go.mod (hunk 1 above)
Comment (richardpark-msft) [reply go.mod:21]: Definitely - will remove it once those PRs merge.

File: rpc.go [hunk 3]
```diff
@@ -39,49 +39,76 @@ import (
...
-		return err
+		return nil, nil, err
 	}
-	r.cancelAuthRefresh, err = r.ec.Namespace().negotiateClaim(ctx, r.client, r.ec.ManagementPath())
+
+	cancelAuthRefresh, err := ec.Namespace().negotiateClaim(ctx, client, ec.ManagementPath())
+
 	if err != nil {
-		return err
+		// it's important to return the `client` here so the calling code
```
Comment (jhendrixMSFT) [inline rpc.go:106]: I think we should just close the client here.  The idiom is, whenever the returned error != `nil` you shouldn't touch the other things being returned.

File: rpc.go (hunk 2 above)
Comment (richardpark-msft) [reply rpc.go:75]: 🤦 

Yes.

File: rpc.go (hunk 3 above)
Comment (richardpark-msft) [reply rpc.go:106]: Good point, will change.

File: rpc.go (hunk 2 above)
Comment (richardpark-msft) [reply rpc.go:75]: I remember why I did this now.

r.client isn't defined at this point (ie, this needs to be lazy evaluated to grab the latest value of `r.client`)

Comment (jhendrixMSFT) [reply rpc.go:75]: Ah ok makes sense.

File: rpc.go [hunk 4]
```diff
@@ -111,7 +142,60 @@ func (r *rpcClient) close() error {
 	if r.cancelAuthRefresh != nil {
 		<-r.cancelAuthRefresh()
 	}
-	return r.client.Close()
+
+	// we don't want to interrupt the cleanup of these links.
+	r.resetLinkCache(context.Background())
```
Comment (jhendrixMSFT) [inline rpc.go:147]: Do callers of `close()` have a context?  If so we should flow it through to here instead of using background.

Comment (richardpark-msft) [reply rpc.go:147]: Unfortunately, no. 

The 'public' function is 'Close()', no context. I promise we won't do that in track 2. :)

File: rpc.go [hunk 5]
```diff
@@ -127,19 +211,35 @@ func (r *rpcClient) doRPCWithRetry(ctx context.Context, address string, msg *amq
 		var link *rpc.Link
 		var rsp *rpc.Response
 		var err error
-		link, err = rpc.NewLink(client, address, opts...)
+		isCachedLink := false
+
+		if len(opts) == 0 {
+			// Can use a cached client since there's nothing unique about this particular
+			// link for this call.
+			// It looks like creating a management link with a session filter is the
```
Comment (jhendrixMSFT) [inline rpc.go:219]: Interesting.  Can this link also be cached?

File: rpc.go (hunk 3 above)
Comment (richardpark-msft) [reply rpc.go]: Changed. It's annoying how I can't test it currently (`Namespace()` is publicly exposed through the Subscription and Queue structs because of embedding).

File: rpc.go (hunk 5 above)
Comment (richardpark-msft) [reply rpc.go:216]: It's trickier than what I was doing for non-sessions because the session filter is targeted to a specific session (ie, not just generic to the entire entity).

I need to do a little more homework for this particular one but I figured defaulting to the old behavior would be fine for now.

Comment (jhendrixMSFT) [reply rpc.go:216]: Makes sense.

File: go.mod (hunk 1 above)
Comment (richardpark-msft) [reply go.mod]: Removed (everything's been published now)

**Review by jhendrixMSFT — APPROVED** [review]

**Comment by richardpark-msft** [comment]
(last live test failure was due to throttling)
